src/codemirror/editor.ts

```diff
@@ -0,0 +1,301 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import * as CodeMirror
+  from 'codemirror';
+
+import 'codemirror/mode/meta';
+
+import {
+  ISignal, defineSignal
+} from 'phosphor/lib/core/signaling';
+
+import {
+  Widget
+} from 'phosphor/lib/ui/widget';
+
+import {
+  AbstractCodeEditor, IPosition, IDimension, IEditorModel, IEditorConfiguration
+} from '../codeeditor/editor';
+
+import {
+  loadModeByFileName, loadModeByMIME
+} from '../codemirror';
+
+/**
+ * The class name added to CodeMirrorWidget instances.
+ */
+const EDITOR_CLASS = 'jp-CodeMirrorWidget';
+
+/**
+ * The name of the default CodeMirror theme
+ */
+export
+const DEFAULT_CODEMIRROR_THEME = 'jupyter';
+
+/**
+ * CodeMirror editor.
+ */
+export
+class CodeMirrorEditor extends AbstractCodeEditor implements IEditorModel, IEditorConfiguration {
+
+  /**
+   * A signal emitted when a uri of this model changed.
+   */
+  uriChanged: ISignal<IEditorModel, void>;
+
+  /**
+   * A signal emitted when a content of this model changed.
+   */
+  contentChanged: ISignal<IEditorModel, void>;
+
+  /**
+   * A signal emitted when a mime type of this model changed.
+   */
+  mimeTypeChanged: ISignal<IEditorModel, void>;
+
+  /**
+   * A signal emitted when this configuration changed.
+   */
+  configurationChanged: ISignal<IEditorConfiguration, void>;
+
+  /**
+   * Construct a CodeMirror editor.
+   */
+  constructor(widget:Widget, options: CodeMirror.EditorConfiguration = {}) {
+    super(widget);
+    this.widget.addClass(EDITOR_CLASS);
+    options.theme = (options.theme || DEFAULT_CODEMIRROR_THEME);
+    this.codeMirrorEditor = CodeMirror(this.widget.node, options);
+
+    const doc = this.codeMirrorEditor.getDoc();
+    CodeMirror.on(doc, 'change', (instance, change) => {
+      this.onDocChange(instance, change);
+    });
+  }
+
+  /**
+   * An underyling CodeMirror editor.
+   */
+  get codeMirrorEditor() {
```

Comment:
Do we want this to be a public property - seems like that it likely to tie other parts of the code base to codemirror...maybe needed though...